syntax = "proto3";

//**************************************************************************
// TR-451 vOMCI SBI Message Protocol Buffer Schema
//
//  Copyright (c) 2021, Broadband Forum
//
//  Redistribution and use in source and binary forms, with or
//   without modification, are permitted provided that the following
//   conditions are met:
//
//   1. Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//   2. Redistributions in binary form must reproduce the above
//      copyright notice, this list of conditions and the following
//      disclaimer in the documentation and/or other materials
//      provided with the distribution.
//
//   3. Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products
//      derived from this software without specific prior written
//      permission.
//
//   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
//   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
//   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
//   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
//   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
//   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
//   NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
//   LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//   CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
//   STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
//   ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//   The above license is used as a license under copyright only.
//   Please reference the Forum IPR Policy for patent licensing terms
//   <https://www.broadband-forum.org/ipr-policy>.
//
//   Any moral rights which are necessary to exercise under the above
//   license grant are also deemed granted under this license.
//
// | Version           | Name                   | Date       |
// | TR-451 1.0.0      | vOMCI Specification    | TBD, 2021  |
//
// BBF software release registry: http://www.broadband-forum.org/software
//**************************************************************************

package tr451_vomci_sbi_message.v1;

message Hello {
  string endpoint_name = 1; //vOMCI function or OLT name
}

message HelloVomciRequest {
  Hello local_endpoint_hello = 1;
}

message HelloVomciResponse {
  Hello remote_endpoint_hello = 1;
}

message VomciMessage {
  string   request_payload_id   = 1; //Optional: The request payload identifier received from the vOLTMF by the vOMCI function.
  oneof msg{
    VomciError error_msg = 2;
    OmciPacket omci_packet_msg = 3;
  }
}


message VomciError {
  enum VomciErrorCode  {
    ERROR_GENERAL = 0;
    NO_ROUTE= 1; //The entity does know which remote endpoint to send the message
    REQUEST_TIMEOUT = 2; //The entity did not receive an expected response for a request
  }
  VomciErrorCode error_code = 1;
  string error_description = 2; //Optional: Error description
}


message OnuHeader {
  string   olt_name = 1; //The OLT name
  string   chnl_term_name = 2; //The reference identifier to the channel termination of the TR-385 vANI
  uint32   onu_id = 3; //The TC layer ONU-ID identifier of the TR-385 vANI
}

message OmciMsgRetrans {
  uint32 t_maxi = 1; //Optional: The maximum threshold value, in milliseconds, for timer expiration as defined in ITU G.988/B.2. A value of 0 indicates that the field is not defined.
  uint32 r_maxi = 2; //Optional: The maximum retries associated with this message as defined in ITU G.988/B.2 plus 1 (e.g., Value 1 is the first retry) A value of 0 indicates that the field is not defined.
}

message MessageInfo {
  OmciMsgRetrans omci_msg_retrans = 1; //Optional: OMCI message retransmission applied
}


message OmciPacket {
  OnuHeader header  = 1;
  MessageInfo msg_info = 2; //Optional: Message information
  bytes    payload   = 3; //OMCI message payload without the CRC/MIC
}


